# <font color='blue'>Data Science Academy - Python Fundamentos - Capítulo 7</font>

## Download: http://github.com/dsacademybr

## Missão: Analisar o Comportamento de Compra de Consumidores.

## Nível de Dificuldade: Alto

Você recebeu a tarefa de analisar os dados de compras de um web site! Os dados estão no formato JSON e disponíveis junto com este notebook.

No site, cada usuário efetua login usando sua conta pessoal e pode adquirir produtos à medida que navega pela lista de produtos oferecidos. Cada produto possui um valor de venda. Dados de idade e sexo de cada usuário foram coletados e estão fornecidos no arquivo JSON.

Seu trabalho é entregar uma análise de comportamento de compra dos consumidores. Esse é um tipo de atividade comum realizado por Cientistas de Dados e o resultado deste trabalho pode ser usado, por exemplo, para alimentar um modelo de Machine Learning e fazer previsões sobre comportamentos futuros.

Mas nesta missão você vai analisar o comportamento de compra dos consumidores usando o pacote Pandas da linguagem Python e seu relatório final deve incluir cada um dos seguintes itens:

** Contagem de Consumidores **

* Número total de consumidores


** Análise Geral de Compras **

* Número de itens exclusivos
* Preço médio de compra
* Número total de compras
* Rendimento total


** Informações Demográficas Por Gênero **

* Porcentagem e contagem de compradores masculinos
* Porcentagem e contagem de compradores do sexo feminino
* Porcentagem e contagem de outros / não divulgados


** Análise de Compras Por Gênero **

* Número de compras
* Preço médio de compra
* Valor Total de Compra
* Compras for faixa etária


** Identifique os 5 principais compradores pelo valor total de compra e, em seguida, liste (em uma tabela): **

* Login
* Número de compras
* Preço médio de compra
* Valor Total de Compra
* Itens mais populares


** Identifique os 5 itens mais populares por contagem de compras e, em seguida, liste (em uma tabela): **

* ID do item
* Nome do item
* Número de compras
* Preço do item
* Valor Total de Compra
* Itens mais lucrativos


** Identifique os 5 itens mais lucrativos pelo valor total de compra e, em seguida, liste (em uma tabela): **

* ID do item
* Nome do item
* Número de compras
* Preço do item
* Valor Total de Compra


** Como considerações finais: **

* Seu script deve funcionar para o conjunto de dados fornecido.
* Você deve usar a Biblioteca Pandas e o Jupyter Notebook.


In [1]:
# Imports
import pandas as pd
import numpy as np

In [2]:
# Carrega o arquivo
load_file = "dados_compras.json"
data = pd.read_json(load_file, orient = "records")
data.head(15)


Login  Idade       Sexo  Item ID  \
0         Aelalis34     38  Masculino      165   
1            Eolo46     21  Masculino      119   
2       Assastnya25     34  Masculino      174   
3      Pheusrical25     21  Masculino       92   
4            Aela59     23  Masculino       63   
5        Tanimnya91     20  Masculino       10   
6       Undjaskla97     20  Masculino      153   
7      Iathenudil29     29   Feminino      169   
8      Sondenasta63     25  Masculino      118   
9        Hilaerin92     31  Masculino       99   
10       Chamosia29     24  Masculino       57   
11          Sally64     20  Masculino       47   
12        Iskossa88     30  Masculino       81   
13  Seorithstilis90     23  Masculino       76   
14        Sundast29     40  Masculino       44   

                                 Nome do Item  Valor  
0                 Bone Crushing Silver Skewer   3.37  
1   Stormbringer, Dark Blade of Ending Misery   2.32  
2                             Primitive Blade   2.46  
3                                Final Critic   1.36  
4                              Stormfury Mace   1.27  
5                                 Sleepwalker   1.73  
6                             Mercenary Sabre   4.57  
7      Interrogator, Blood Blade of the Queen   3.32  
8            Ghost Reaver, Longsword of Magic   2.77  
9        Expiration, Warscythe Of Lost Worlds   4.53  
10            Despair, Favor of Due Diligence   3.81  
11                Alpha, Reach of Ending Hope   1.55  
12                                  Dreamkiss   4.06  
13                 Piety, Guardian of Riddles   3.68  
14                      Bonecarvin Battle Axe   2.46

In [3]:
display(data)

Login  Idade       Sexo  Item ID  \
0       Aelalis34     38  Masculino      165   
1          Eolo46     21  Masculino      119   
2     Assastnya25     34  Masculino      174   
3    Pheusrical25     21  Masculino       92   
4          Aela59     23  Masculino       63   
..            ...    ...        ...      ...   
775       Eural50     22  Masculino       98   
776    Lirtossa78     14  Masculino      104   
777    Tillyrin30     20  Masculino      117   
778    Quelaton80     20  Masculino       75   
779        Alim85     23   Feminino      107   

                                  Nome do Item  Valor  
0                  Bone Crushing Silver Skewer   3.37  
1    Stormbringer, Dark Blade of Ending Misery   2.32  
2                              Primitive Blade   2.46  
3                                 Final Critic   1.36  
4                               Stormfury Mace   1.27  
..                                         ...    ...  
775                Deadline, Voice Of Subtlety   3.62  
776                         Gladiator's Glaive   1.36  
777          Heartstriker, Legacy of the Light   4.15  
778                    Brutality Ivory Warmace   1.72  
779                  Splitter, Foe Of Subtlety   3.61  

[780 rows x 6 columns]

## Informações Sobre os Consumidores

In [4]:
# Número total de consumidores
data.groupby(['Login'], dropna=True).count().shape


(573, 5)

A tabela informa que existem 573 consumidores cadastrados

## Análise Geral de Compras

In [5]:
# Número de itens exclusivos
data.groupby('Item ID').count().shape

(183, 5)

A tabela informa que existem 183 itens cadastexclusivos

In [6]:
# Preço médio de compra
print("Valor :", data['Valor'].mean())


Valor : 2.931192307692303


In [7]:
# Número total de compras
print("Quantidade total de compras",data[data.columns[0]].count())

Quantidade total de compras 780


In [8]:
# Rendimento total
print("Rendimento total", data['Valor'].sum())

Rendimento total 2286.33


## Informações Demográficas Por Gênero

In [149]:
# Porcentagem e contagem de compradores masculinos
# Porcentagem e contagem de compradores do sexo feminino
# Porcentagem e contagem de outros / não divulgados
valores = data.groupby('Sexo').sum()
valores['Percent'] = (valores['Valor'] * 100) / valores['Valor'].sum()
valores['Percent'] = valores['Percent'].map("%{:,.2f}".format)


In [150]:
display(valores)

Idade  Item ID    Valor Percent
Sexo                                                  
Feminino                3068    11983   382.91  %16.75
Masculino              14360    57964  1867.68  %81.69
Outro / Não Divulgado    301     1261    35.74   %1.56

Criada uma coluna com o nome "Percent" indicando as porcentagens de cada genero

## Análise de Compras Por Gênero

In [73]:
# Número de compras
# Preço médio de compra

valores = data.groupby('Sexo')['Sexo'].count()
print('Tabela com o total de compras por genero \n \n',valores)

Tabela com o total de compras por genero 
 
 Sexo
Feminino                 136
Masculino                633
Outro / Não Divulgado     11
Name: Sexo, dtype: int64


In [76]:
# Valor Total de Compra

valores = data.groupby('Sexo')['Valor'].sum()
print('Tabela com o valor total gasto por genero \n \n', valores)

Tabela com o valor total gasto por genero 
 
 Sexo
Feminino                  382.91
Masculino                1867.68
Outro / Não Divulgado      35.74
Name: Valor, dtype: float64


In [78]:
# Compras for faixa etária
# Criando uma coluna indicando a faixa etaria 
faixa = [0, 17, 29, 40, 65, 120]
labels = ['Menor de Idade', 'Jovem', 'Adulto', 'idoso', 'Melhor Idade']
faixa = pd.cut(x=data.Idade, bins=faixa, labels=labels)

In [79]:
# Adcionando a coluna a tabela original
data['Faixa Etaria'] = faixa
data.head()

Login  Idade       Sexo  Item ID  \
0     Aelalis34     38  Masculino      165   
1        Eolo46     21  Masculino      119   
2   Assastnya25     34  Masculino      174   
3  Pheusrical25     21  Masculino       92   
4        Aela59     23  Masculino       63   

                                Nome do Item  Valor Faixa Etaria  
0                Bone Crushing Silver Skewer   3.37       Adulto  
1  Stormbringer, Dark Blade of Ending Misery   2.32        Jovem  
2                            Primitive Blade   2.46       Adulto  
3                               Final Critic   1.36        Jovem  
4                             Stormfury Mace   1.27        Jovem

In [80]:
valores = data.groupby('Faixa Etaria')['Valor'].sum()
print('Tabela com o valor total gasto por Faixa Etaria \n \n', valores)

Tabela com o valor total gasto por genero 
 
 Faixa Etaria
Menor de Idade     432.59
Jovem             1483.34
Adulto             361.76
idoso                8.64
Melhor Idade         0.00
Name: Valor, dtype: float64


## Consumidores Mais Populares (Top 5)

In [97]:
valores = data.groupby('Login')['Login'].count().sort_values(ascending=False)
print('Tabela com os Top 5 maiores consumidores \n \n', valores.head(5))

Tabela com os Top 5 maiores consumidores 
 
 Login
Undirrala66    5
Mindimnya67    4
Qarwen67       4
Saedue76       4
Sondastan54    4
Name: Login, dtype: int64


## Itens Mais Populares

In [109]:
valores = data.groupby('Nome do Item')['Nome do Item'].count().sort_values(ascending=False)
print('Tabela com itens mais populares \n \n', valores.head(10))

Tabela com itens mais populares 
 
 Nome do Item
Final Critic                            14
Arcane Gem                              11
Betrayal, Whisper of Grieving Widows    11
Stormcaller                             10
Woeful Adamantite Claymore               9
Serenity                                 9
Trickster                                9
Retribution Axe                          9
Splitter, Foe Of Subtlety                8
Darkheart                                8
Name: Nome do Item, dtype: int64


## Itens Mais Lucrativos

In [151]:
total_valor = data.groupby("Nome do Item").sum()["Valor"].rename("Valor Total")
media_valor = data.groupby("Nome do Item").mean()["Valor"].rename("Valor Total")
quant_venda = data.groupby("Nome do Item").count()["Valor"].rename("Valor Total")

# Dataframe para os resultados
itens = pd.DataFrame({"Valor Total": total_valor,
                      "Valor Medio": media_valor,
                      "Quantidade": quant_venda})

itens["Valor Total"] = itens["Valor Total"].map("${:,.2f}".format)
itens["Valor Medio"] = itens["Valor Medio"].map("${:,.2f}".format)

itens.head(10).sort_values(by='Quantidade',ascending=False)

Valor Total Valor Medio  Quantidade
Nome do Item                                                      
Arcane Gem                          $24.53       $2.23          11
Alpha                               $10.92       $1.56           7
Alpha, Oath of Zeal                 $20.16       $2.88           7
Alpha, Reach of Ending Hope          $9.30       $1.55           6
Amnesia                             $21.42       $3.57           6
Avenger                             $24.96       $4.16           6
Agatha                               $9.55       $1.91           5
Aetherius, Boon of the Blessed      $19.00       $4.75           4
Apocalyptic Battlescythe            $15.64       $3.91           4
Abyssal Shard                        $6.12       $2.04           3

## Fim

### Obrigado

### Visite o Blog da Data Science Academy - <a href="http://blog.dsacademy.com.br">Blog DSA</a>